src/rust/catnap/queue.rs

```diff
@@ -6,85 +6,220 @@
 //======================================================================================================================
 
 use crate::{
-    catnap::socket::Socket,
+    catnap::socket::{
+        retry_errno,
+        Socket,
+    },
     runtime::{
         fail::Fail,
+        limits,
+        memory::DemiBuffer,
         queue::{
             IoQueue,
             QType,
         },
     },
     scheduler::{
         TaskHandle,
+        Yielder,
         YielderHandle,
     },
 };
 use ::std::{
+    cell::{
+        Ref,
+        RefCell,
+        RefMut,
+    },
     collections::HashMap,
-    os::unix::prelude::RawFd,
+    net::SocketAddrV4,
+    rc::Rc,
 };
 
 //======================================================================================================================
 // Structures
 //======================================================================================================================
 
 /// Per-queue metadata: Catnap control block
+#[derive(Clone)]
 pub struct CatnapQueue {
     qtype: QType,
-    fd: Option<RawFd>,
-    socket: Socket,
-    pending_ops: HashMap<TaskHandle, YielderHandle>,
+    socket: Rc<RefCell<Socket>>,
+    pending_ops: Rc<RefCell<HashMap<TaskHandle, YielderHandle>>>,
 }
 
 //======================================================================================================================
 // Associated Functions
 //======================================================================================================================
 
 impl CatnapQueue {
-    pub fn new(qtype: QType, fd: Option<RawFd>) -> Self {
+    pub fn new(qtype: QType, socket: Socket) -> Self {
         Self {
             qtype,
-            fd,
-            socket: Socket::new(),
-            pending_ops: HashMap::<TaskHandle, YielderHandle>::new(),
+            socket: Rc::new(RefCell::new(socket)),
+            pending_ops: Rc::new(RefCell::new(HashMap::<TaskHandle, YielderHandle>::new())),
+        }
+    }
+
+    pub fn local(&self) -> Option<SocketAddrV4> {
+        self.socket.borrow().local()
+    }
+
+    pub fn remote(&self) -> Option<SocketAddrV4> {
+        self.socket.borrow().remote()
+    }
+
+    pub fn bind(&mut self, local: SocketAddrV4) -> Result<(), Fail> {
+        self.socket.borrow_mut().bind(local)
+    }
+
+    pub fn listen(&mut self, backlog: usize) -> Result<(), Fail> {
+        self.socket.borrow_mut().listen(backlog)
+    }
+
+    pub fn set_accept(&mut self) -> Result<(), Fail> {
+        self.socket.borrow_mut().accept()
+    }
+
+    pub async fn accept(&mut self, yielder: Yielder) -> Result<Self, Fail> {
+        loop {
+            let mut socket = self.socket.borrow_mut();
+            match socket.try_accept() {
+                Ok(socket) => {
```

Comment:
Would it be better to name this `accepted_socket` instead, to avoid confusion?

Reply:
Yup, renamed this new_accepted_socket